lib/spack/spack/spec.py

```diff
@@ -1267,11 +1267,17 @@ def cshort_spec(self):
 
     @property
     def prefix(self):
+        assert self._concrete
         if self._prefix is None:
             upstream, record = spack.store.db.query_by_spec_hash(
                 self.dag_hash())
             if record and record.path:
-                self.prefix = record.path
+                # Old databases may have 'None' (the string not the value)
+                # for paths for externals. Use external_path for the prefix.
+                if record.path != 'None':
+                    self.prefix = record.path
+                else:
+                    self.prefix = record.spec.external_path
```

Comment:
This is the wrong place for this -- issues with old database formats should be dealt with on read-in, not in the other methods in `spec.py`. 

make it so if the reader encounters an old external with `"None"` as the path value, that it converts that to the modern spec format.  Then other methods (like `prefix`) don't have to think about historical anomalies.